File: babbleon/storage.py
"""
On-disk paths for vault, state, config.

Honors XDG_CONFIG_HOME / XDG_DATA_HOME. M1 uses ~/.config/babbleon and
~/.local/share/babbleon by default.
"""
import os
import pathlib


def config_dir() -> pathlib.Path:
    base = os.environ.get("XDG_CONFIG_HOME") or "~/.config"
    return pathlib.Path(base).expanduser() / "babbleon"


def data_dir() -> pathlib.Path:
    base = os.environ.get("XDG_DATA_HOME") or "~/.local/share"
    return pathlib.Path(base).expanduser() / "babbleon"


def vault_path() -> pathlib.Path:
    return config_dir() / "vault.age"


def state_path() -> pathlib.Path:
    return config_dir() / "state.json"


def ensure_dirs() -> None:
    config_dir().mkdir(parents=True, exist_ok=True)
    data_dir().mkdir(parents=True, exist_ok=True)
    try:
        config_dir().chmod(0o700)
    except PermissionError:
        pass

File: crates/babbleon/src/storage.rs
//! XDG-aware storage paths.

use std::path::PathBuf;

pub fn config_dir() -> PathBuf {
    dirs::config_dir()
        .unwrap_or_else(|| PathBuf::from("/tmp"))
        .join("babbleon")
}

pub fn data_dir() -> PathBuf {
    dirs::data_dir()
        .unwrap_or_else(|| PathBuf::from("/tmp"))
        .join("babbleon")
}

pub fn vault_path() -> PathBuf {
    config_dir().join("vault.age")
}

pub fn state_path() -> PathBuf {
    config_dir().join("state.json")
}

pub fn ensure_dirs() -> std::io::Result<()> {
    std::fs::create_dir_all(config_dir())?;
    std::fs::create_dir_all(data_dir())?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let _ = std::fs::set_permissions(config_dir(), std::fs::Permissions::from_mode(0o700));
    }
    Ok(())
}
